### PR TITLE
docs: Improve the distributed training guide

### PR DESCRIPTION
**Pull request by tara-det-ai**
This is an effort to continue improving the distributing training guide. For related PR with additional comments, visit https://github.com/determined-ai/determined/pull/6983.

## Checklist

- [x] Meets documentation standards and style guidelines.
- [x] Content is clear, correct, consistent.

### CODE REVIEW

File: docs/model-dev-guide/dtrain/dtrain-implement.rst [hunk 1]
```diff
@@ -24,110 +25,117 @@ training, as described in :ref:`cluster-configuration`.
...
-shared by the agent machines. If your cluster has multiple common network interfaces, please specify
-the fastest one in :ref:`cluster-configuration` under
+For distributed multi-machine training, Determined will automatically detect a common network
+interface that is shared by the agent machines. If your cluster has multiple common network
+interfaces, we advise specifying the fastest one in :ref:`cluster-configuration` under
 ``task_container_defaults.dtrain_network_interface``.
 
-Reduce computational overhead by setting the ``global_batch_size`` to the largest batch size that
-fits into a single GPU multiplied times the number of slots.
+YOu can reduce computational overhead by setting the ``global_batch_size`` to the largest batch size
```
Comment (dpsi4) [inline docs/model-dev-guide/dtrain/dtrain-implement.rst:57]: Double capitalization on "YOu" at the start of this sentence.

File: docs/model-dev-guide/dtrain/dtrain-implement.rst [hunk 2]
```diff
@@ -24,110 +25,117 @@ training, as described in :ref:`cluster-configuration`.
...
+loading. You're encouraged to experiment with a synthetic dataset in order to verify the performance
+of multi-GPU training.
 
 .. warning::
 
    Multi-machine distributed training is designed to maximize performance by training with all the
    resources of a machine. This can lead to situations where an experiment is created but never
-   becomes active: if the number of GPUs requested does not divide into the machines available, for
-   instance, or if another experiment is already using some GPUs on a machine.
+   becomes active such as when the number of GPUs requested does not divide into the machines
```
Comment (dpsi4) [inline docs/model-dev-guide/dtrain/dtrain-implement.rst:135]: There should be a comma in here "becomes active, such as...."

Also, I would slightly rephrase this as "...number of GPUs does not divide **evenly** into the machines..." (though not bold, that's just to make the suggestion stand out for you)

**Review by dpsi4 — COMMENTED** [review]
I added a couple of comments on minor issues, but otherwise it looks good.

File: docs/model-dev-guide/dtrain/dtrain-implement.rst [hunk 3]
```diff
@@ -24,110 +25,117 @@ training, as described in :ref:`cluster-configuration`.
...
 
-If :ref:`slots_per_trial <exp-config-resources-slots-per-trial>` is greater than the number of slots
-on a single agent, Determined schedules it over multiple machines. When scheduling a multi-machine
-distributed training job, Determined requires that the job uses all of the slots (GPUs) on an agent.
-For example, in a cluster that consists of 8-GPU agents, an experiment with :ref:`slots_per_trial
-<exp-config-resources-slots-per-trial>` set to ``12`` is never scheduled and will wait indefinitely.
-The section on :ref:`Scheduling Behavior <dtrain-scheduling>` describes this in more detail.
+If the value of :ref:`slots_per_trial <exp-config-resources-slots-per-trial>` is greater than the
+number of slots available on a single agent, Determined schedules it over multiple machines. When
+scheduling a multi-machine distributed training job, Determined requires that all slots (GPUs) on an
```
Comment (azhou-determined) [inline docs/model-dev-guide/dtrain/dtrain-implement.rst:69]: should this info about uneven slots be a separate note? not sure what our standard is for note/warning sections. but seems like something worth calling out, slightly unexpected behavior.

File: docs/model-dev-guide/dtrain/dtrain-implement.rst [hunk 4]
```diff
@@ -24,110 +25,117 @@ training, as described in :ref:`cluster-configuration`.
...
+The ``resources.slots_per_trial`` field in the :ref:`experiment-config-reference` controls the
 number of GPUs used to train a single trial.
 
-The default value is ``1``, which disables distributed training. Setting ``slots_per_trial`` to a
-larger value enables multi-GPU training automatically. Note that these GPUs might be on a single
-machine or across multiple machines; the experiment configuration simply defines how many GPUs
-should be used for training, and the Determined job scheduler decides whether to schedule the task
-on a single agent or multiple agents, depending on the machines in the cluster and the other active
-workloads.
+By default, this field is set to a value of ``1``, which essentially disables distributed training.
```
Comment (azhou-determined) [inline docs/model-dev-guide/dtrain/dtrain-implement.rst:31]: nit: essentially -> effectively?

File: docs/model-dev-guide/dtrain/dtrain-implement.rst [hunk 5]
```diff
@@ -24,110 +25,117 @@ training, as described in :ref:`cluster-configuration`.
...
-shared by the agent machines. If your cluster has multiple common network interfaces, please specify
-the fastest one in :ref:`cluster-configuration` under
+For distributed multi-machine training, Determined will automatically detect a common network
+interface that is shared by the agent machines. If your cluster has multiple common network
+interfaces, we advise specifying the fastest one in :ref:`cluster-configuration` under
 ``task_container_defaults.dtrain_network_interface``.
 
-Reduce computational overhead by setting the ``global_batch_size`` to the largest batch size that
-fits into a single GPU multiplied times the number of slots.
+You can reduce computational overhead by setting the ``global_batch_size`` to the largest batch size
```
Comment (azhou-determined) [inline docs/model-dev-guide/dtrain/dtrain-implement.rst:57]: do we need this sentence here since we already have a section right below this for global batch size?

File: docs/model-dev-guide/dtrain/dtrain-introduction.rst [hunk 6]
```diff
@@ -8,56 +8,56 @@
...
 #. Every worker averages the updates by the number of workers.
 #. Every worker applies the updates to its copy of the model parameters, resulting in all the
    workers having identical solution states.
 #. Return to the first step.
 
-***********************************************
- Reduce Computation and Communication Overhead
-***********************************************
+*************************************************
+ Reducing Computation and Communication Overhead
```
Comment (azhou-determined) [inline docs/model-dev-guide/dtrain/dtrain-introduction.rst:29]: should this section belong in `Debug Performance Bottlenecks`? it's a bit confusing to me how this document is organized. feel like i have to read the whole thing, each section doesn't necessarily cleanly encapsulate its info.

File: docs/model-dev-guide/dtrain/dtrain-implement.rst (hunk 4 above)
Comment (garrett361) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:31]: Or just "which disables"?

File: docs/model-dev-guide/dtrain/dtrain-implement.rst [hunk 7]
```diff
@@ -24,110 +25,117 @@ training, as described in :ref:`cluster-configuration`.
...
 Global Batch Size
 =================
 
-When doing distributed training, the ``global_batch_size`` specified in the
+During distributed training, the ``global_batch_size`` specified in the
 :ref:`experiment-config-reference` is partitioned across ``slots_per_trial`` GPUs. The per-GPU batch
 size is set to: ``global_batch_size // slots_per_trial``. If ``slots_per_trial`` does not divide
 ``global_batch_size`` evenly, the remainder is dropped. For convenience, the per-GPU batch size can
 be accessed via the Trial API, using :func:`context.get_per_slot_batch_size
 <determined.TrialContext.get_per_slot_batch_size>`.
```
Comment (garrett361) [inline docs/model-dev-guide/dtrain/dtrain-implement.rst:87]: We probably should insert an explicit caveat somewhere that this is a Trial API-only feature and that with Core API the user needs to handle batch size configuration manually.

File: docs/model-dev-guide/dtrain/dtrain-implement.rst (hunk 3 above)
Comment (garrett361) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:69]: @azhou-determined is the uneven slot constraint still true? I thought we changed the scheduler behavior so that you could use uneven slots. Or at least had some experimental feature that could do so.

I also think this paragraph is a little redundant with the earlier paragraph about multi-machine parallelism. Or at least the two paragraphs could be merged.

File: docs/model-dev-guide/dtrain/dtrain-implement.rst [hunk 8]
```diff
@@ -24,110 +25,117 @@ training, as described in :ref:`cluster-configuration`.
...
 
-For improved performance, *weak-scaling* is recommended. That is, increasing your
-``global_batch_size`` proportionally with ``slots_per_trial``. For example, change
-``global_batch_size`` and ``slots_per_trial`` from 32 and 1 to 128 and 4.
+For improved performance, *weak-scaling* is recommended. Weak-scaling means proportionally
```
Comment (garrett361) [inline docs/model-dev-guide/dtrain/dtrain-implement.rst:89]: nit: seems annoyingly technical to introduce jargon (weak-scaling) that is only used in this immediate sentence and without any link that could explain the concept further.

File: docs/model-dev-guide/dtrain/dtrain-implement.rst [hunk 9]
```diff
@@ -136,17 +144,19 @@ multi-GPU training.
  Downloading Data
```
Comment (garrett361) [inline docs/model-dev-guide/dtrain/dtrain-implement.rst:144]: This seems like sub-optimal advice: don't we only need to download the data once per _node_, not once per gpu?

File: docs/model-dev-guide/dtrain/dtrain-implement.rst [hunk 10]
```diff
@@ -196,19 +205,19 @@ distributed training:
    distributed training jobs are not scheduled and wait indefinitely. For example, if every agent in
    the cluster has eight GPUs, a job with ``slots_per_trial`` set to ``12`` is never scheduled.
 
-   If a multi-GPU experiment does not become active after a minute or so, please confirm that
-   ``slots_per_trial`` is set so that it can be scheduled within these constraints. The CLI command
-   ``det task list`` can also be used to check if any other tasks are using GPUs and preventing your
+   If a multi-GPU experiment does not become active after a minute or so, please ensure that
+   ``slots_per_trial`` is set so that it can be scheduled within these constraints. You can also use
+   the CLI command ``det task list`` to check if any other tasks are using GPUs and preventing your
```
Comment (garrett361) [inline docs/model-dev-guide/dtrain/dtrain-implement.rst:210]: Redundant, again.

File: docs/model-dev-guide/dtrain/dtrain-implement.rst [hunk 11]
```diff
@@ -219,40 +228,38 @@ batch inference job, create a new PyTorchTrial and follow these steps:
 Once the new PyTorchTrial object is created, use the experiment configuration to distribute
 inference in the same way as training. `cifar10_pytorch_inference
 <https://github.com/determined-ai/determined/blob/master/examples/computer_vision/cifar10_pytorch_inference/>`_
-is an example of distributed batch inference.
+serves as an example of distributed batch inference.
 
 .. _config-template:
 
 *************************
  Configuration Templates
```
Comment (garrett361) [inline docs/model-dev-guide/dtrain/dtrain-implement.rst:236]: The entire **Configuration Templates** sections seems misplaced on this page. It is not distributed-training-specific advice (though it is helpful information, which I think deserves its own page).

File: docs/model-dev-guide/dtrain/dtrain-introduction.rst [hunk 12]
```diff
@@ -8,93 +8,93 @@
...
 *******************************************
 
-Distributed training in Determined utilizes data-parallelism. Data-parallelism for deep-learning
-consists of a set of workers, where each worker is assigned to a unique compute accelerator such as
-a GPU or a TPU. Each worker maintains a copy of the model parameters (weights that are being
-trained), which is synchronized across all the workers at the start of training.
+Determined employs data parallelism in its approach to distributed training. Data parallelism for
+deep learning consists of a set of workers, where each worker is assigned to a unique compute
+accelerator such as a GPU or a TPU. Each worker maintains a copy of the model parameters (weights
+that are being trained), which is synchronized across all the workers at the start of training.
```
Comment (garrett361) [inline docs/model-dev-guide/dtrain/dtrain-introduction.rst:14]: Not sure it's worth noting in this document, but this is only one of several parallelism strategies that the user can employ.

File: docs/model-dev-guide/dtrain/dtrain-introduction.rst [hunk 13]
```diff
@@ -8,93 +8,93 @@
...
 
 After initialization is completed, distributed training in Determined follows a loop where:
 
 #. Every worker performs a forward and backward pass on a unique mini-batch of data.
```
Comment (garrett361) [inline docs/model-dev-guide/dtrain/dtrain-introduction.rst:18]: Could consider a diagram here? Would help break up the text.

File: docs/model-dev-guide/dtrain/dtrain-introduction.rst [hunk 14]
```diff
@@ -8,93 +8,93 @@
...
-the computational overhead. To reduce computational overhead, it's recommended that users maximize
-the utilization of their GPU. This is typically done by using the largest batch size that fits into
-memory. When performing distributed training, to reduce the computational overhead it's recommended
-to set the ``global_batch_size`` to the largest batch size that fits into a single GPU multiplied
-times the number of slots. This is commonly referred to as *weak scaling*.
+In the distributed training loop, the first two steps bring about the most computational burden,
+known as computational overhead. To reduce this computational overhead, we recommend using your GPU
+at its maximimum capacity. This is typically achieved by using the largest batch size that fits into
+memory. More specifically, we recommend setting the ``global_batch_size`` to the largest batch size
+that fits into a single GPU multiplied by the number of slots. This is commonly referred to as *weak
```
Comment (garrett361) [inline docs/model-dev-guide/dtrain/dtrain-introduction.rst:36]: nit: I would remove the _weak scaling_ sentence because it's both overly-jargony (imo), but also not accurate in this paragraph.

File: docs/model-dev-guide/dtrain/dtrain-introduction.rst [hunk 15]
```diff
@@ -8,93 +8,93 @@
...
-scaling performance, assuming you have scaled your ``global_batch_size`` proportionally with
-``slots_per_trial``, it's possible that training performance is being bottlenecked by network
-communication or disk I/O.
+Deep learning models typically perform dense updates, meaning every model parameter is updated for
+every training sample. Consequently, the quantity of communication per mini-batch (step 3 in the
+distributed training loop) is dependent on the number of model parameters. Models having fewer
+parameters like `ResNet-50 <https://arxiv.org/pdf/1512.03385.pdf>`_ (~30 million parameters) train
+more efficiently in distributed settings than models with more parameters such as `VGG-16
+<https://arxiv.org/pdf/1505.06798.pdf>`_ (~136 million parameters). If you are planning on utilizing
+distributed training, consider the size of your model when designing it.
```
Comment (garrett361) [inline docs/model-dev-guide/dtrain/dtrain-introduction.rst:86]: Not super-helpful advice, and given the title of the section I would also expect an explanation of multiple characteristics that affect performance, rather than just one.  IMO we could just cut this paragraph/section.

File: docs/model-dev-guide/dtrain/dtrain-implement.rst (hunk 4 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:31]: edited. "which disables"

File: docs/model-dev-guide/dtrain/dtrain-implement.rst
[the remark this answers is not in the document]
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:1]: agree

File: docs/model-dev-guide/dtrain/dtrain-implement.rst (hunk 11 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:236]: done

File: docs/model-dev-guide/dtrain/dtrain-introduction.rst (hunk 13 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-introduction.rst:18]: Agree. I'll gather some requirements for the diagram.

File: docs/model-dev-guide/dtrain/dtrain-implement.rst (hunk 5 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:57]: for discussion: where does the user start thinking about or need to know about reducing computational overhead?

File: docs/model-dev-guide/dtrain/dtrain-implement.rst [hunk 16]
```diff
@@ -24,110 +25,117 @@ training, as described in :ref:`cluster-configuration`.
...
+depends on the machines in the cluster and other active workloads.
 
-Multi-machine parallelism offers the ability to further parallelize training across more GPUs. To
-use multi-machine parallelism, set ``slots_per_trial`` to be a multiple of the total number of GPUs
-on an agent machine. For example, if your resource pool consists of multiple 8-GPU agent machines,
-valid values for ``slots_per_trial`` would be 16, 24, 32, etc. In this configuration, trials use all
-the resources of multiple machines to train a model:
+Multi-machine parallelism allows you to further parallelize training across more GPUs. To use this
+feature, ``slots_per_trial`` should be set as a multiple of the total number of GPUs on an agent
+machine. For example, if your resource pool consists of multiple 8-GPU agent machines, valid
```
[the remark this answers is not in the document]
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:41]: add a note with a link to the relevant parameter

File: docs/model-dev-guide/dtrain/dtrain-implement.rst [hunk 17]
```diff
@@ -24,110 +25,117 @@ training, as described in :ref:`cluster-configuration`.
...
-and/or testing accuracy. You may need to adjust model hyperparameters like the learning rate and/or
-use a different optimizer when training with larger batch sizes.
+Note that adjusting ``global_batch_size`` can impact your model convergence, which in turn can
+affect your training and/or testing accuracy. You might need to adjust model hyperparameters like
+the learning rate, or consider using a different optimizer when training with larger batch sizes.
 
 .. _multi-gpu-training-implement-adv-optimizations:
 
 Advanced Optimizations
 ======================
```
[the remark this answers is not in the document]
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:101]: applies to advanced optimizations section

File: docs/model-dev-guide/dtrain/dtrain-implement.rst (hunk 9 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:144]: @garrett361 to help with rewrite

File: docs/model-dev-guide/dtrain/dtrain-introduction.rst (hunk 12 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-introduction.rst:14]: Determined also uses deep speed, for example
we can ignore this comment

File: docs/model-dev-guide/dtrain/dtrain-introduction.rst (hunk 13 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-introduction.rst:18]: created a Jira ticket and a mockup, likely a separate PR

File: docs/model-dev-guide/dtrain/dtrain-implement.rst (hunk 17 above)
[the remark this answers is not in the document]
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:101]: resolved, added a note to let users know

File: docs/model-dev-guide/dtrain/dtrain-implement.rst
[the remark this answers is not in the document]
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:1]: removed "introduction" from the title

[the remark this answers is not in the document]
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:1]: i'll allow this to be a grab bag for now and attempt some reorganization

File: docs/model-dev-guide/dtrain/dtrain-introduction.rst
[the remark this answers is not in the document]
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-introduction.rst:1]: moved the Reproducibility section into its own page; moved the Optimizing Training section into its own page. they'll live under the dtrain directory for now.

File: docs/model-dev-guide/dtrain/dtrain-introduction.rst (hunk 13 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-introduction.rst:18]: i'll set this to resolved since the graphic design is a separate ticket

File: docs/model-dev-guide/dtrain/dtrain-implement.rst (hunk 3 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:69]: answer:

re: "When scheduling a multi-machine distributed training job, Determined requires that the job uses all of the slots (GPUs) on an agent. For example, in a cluster that consists of 8-GPU agents, an experiment with [slots_per_trial](https://docs.determined.ai/latest/reference/training/experiment-config-reference.html#exp-config-resources-slots-per-trial) set to 12 is never scheduled and will wait indefinitely."

Determined now supports uneven slots, see the Scheduler setting "allow_heterogeneous_fits" documented in the master config doc: https://docs.determined.ai/latest/reference/deploy/config/master-config-reference.html#allow-heterogeneous-fits

Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst]: resolved

File: docs/model-dev-guide/dtrain/dtrain-implement.rst (hunk 5 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst]: resolved, moved the sentence into the section about global batch size

File: docs/model-dev-guide/dtrain/dtrain-introduction.rst (hunk 6 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-introduction.rst:29]: yes this section belongs with debugging performance bottlenecks.

Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-introduction.rst:29]: resolved, added a new top level section, "Performance Optimization" and underneath you'll see the two sections: "Reducing Computation and Communication Overhead" and "Debugging Performance Bottlenecks"

File: docs/model-dev-guide/dtrain/dtrain-implement.rst (hunk 16 above)
[the remark this answers is not in the document]
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:42]: @garrett361 what did you find out?

File: docs/model-dev-guide/dtrain/dtrain-implement.rst (hunk 7 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:86]: @garrett361 there is now a note added to the Advanced Optimizations section, please advise if this note should be in other sections as well.

Note

These optimizations only apply to [High-Level APIs](http://localhost:1234/model-dev-guide/apis-howto/overview.html#high-level-apis) (Trial APIs) and do not apply to the Core API.

File: docs/model-dev-guide/dtrain/dtrain-implement.rst (hunk 9 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:148]: @garrett361 which part do we want to rewrite?

File: docs/model-dev-guide/dtrain/dtrain-implement.rst [hunk 18]
```diff
@@ -172,22 +182,21 @@ do this when downloading data from S3. In this example, the S3 bucket name is co
...
    Determined schedules the distributed training job onto multiple agents. A multi-machine
-   distributed training job is only scheduled onto an agent if this results in utilizing all of the
-   agent GPUs. This is to ensure good performance and utilize the full network bandwidth of each
-   machine while minimizing inter-machine networking. For example, if all of the agents in your
-   cluster have eight GPUs each , you should submit jobs with ``slots_per_trial`` set to a multiple
+   distributed training job is only scheduled onto an agent if it results in utilizing all of the
+   agent GPUs. This strategy ensures good performance and utilizes the full network bandwidth of
+   each machine while minimizing inter-machine networking. For example, if all the agents in your
+   cluster have eight GPUs each, you should submit jobs with ``slots_per_trial`` set to a multiple
    of eight, such as 8, 16, or 24.
```
[the remark this answers is not in the document]
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst]: Resolved. This was resolved by consolidating the information into the Scheduling Behavior section.

File: docs/model-dev-guide/dtrain/dtrain-implement.rst (hunk 10 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:220]: This should be resolved now that the info is consolidated in the Scheduling Behavior section.

File: docs/model-dev-guide/dtrain/dtrain-introduction.rst (hunk 14 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-introduction.rst]: @garrett361 let's revisit this - see the other comment about weak scaling

File: docs/model-dev-guide/dtrain/dtrain-implement.rst (hunk 16 above)
[the remark this answers is not in the document]
Comment (garrett361) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:42]: This was in reference to the `allow_heterogeneous_fits` parameter, so should be resolved now

File: docs/model-dev-guide/dtrain/dtrain-implement.rst (hunk 8 above)
Comment (garrett361) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:88]: If you think a paper link is appropriate, I would suggest [this one](https://arxiv.org/abs/1812.06162). But I was just suggesting removing the phrase `weak-scaling` completely.

If you think it's fine as it is, it can also be left alone. Just a personal preference thing.

File: docs/model-dev-guide/dtrain/dtrain-introduction.rst (hunk 15 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-introduction.rst:87]: ok, removed

File: docs/model-dev-guide/dtrain/dtrain-implement.rst (hunk 9 above)
Comment (garrett361) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:148]: I think it should be more like
```python
import boto3
import os

def download_data_from_s3(self):
   s3_bucket = self.context.get_data_config()["bucket"]
   download_directory = "/tmp/data"
   data_file = "data.csv"
   s3 = boto3.client("s3")
   os.makedirs(download_directory, exist_ok=True)
   filepath = os.path.join(download_directory, data_file)
   if not os.path.exists(filepath):
       s3.download_file(s3_bucket, data_file, filepath)
   return download_directory

if not self.context.distributed.local_rank:
    download_directory = self.download_data_from_s3()
    self.context.distributed.broadcast_local(download_directory)
else:
    download_directory = self.context.distribued.broadcast_local(None)
```

Seem right, @azhou-determined ?


I think we'd also want to re-write some examples which use very similar code also, so can also mark this as out of scope?

Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:148]: i'll mark it as out of scope for this PR and create a new one to improve the examples, i'll transfer this comment over

**Review by azhou-determined — APPROVED** [review]

File: docs/model-dev-guide/dtrain/dtrain-implement.rst (hunk 7 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:86]: i had the note in the wrong section. moved it under Global Batch Size.

File: docs/model-dev-guide/dtrain/dtrain-implement.rst (hunk 8 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:88]: resolved via slack

File: docs/model-dev-guide/dtrain/dtrain-implement.rst [hunk 19]
```diff
@@ -78,20 +78,28 @@ Global Batch Size
...
-During distributed training, the ``global_batch_size`` specified in the
-:ref:`experiment-config-reference` is partitioned across ``slots_per_trial`` GPUs. The per-GPU batch
-size is set to: ``global_batch_size // slots_per_trial``. If ``slots_per_trial`` does not divide
-``global_batch_size`` evenly, the remainder is dropped. For convenience, the per-GPU batch size can
-be accessed via the Trial API, using :func:`context.get_per_slot_batch_size
-<determined.TrialContext.get_per_slot_batch_size>`.
+.. note::
+
+   This feature only applies to :ref:`high-level-apis` (Trial APIs) and does not apply to the Core
+   API.
```
Comment (tara-det-ai) [inline docs/model-dev-guide/dtrain/dtrain-implement.rst:84]: @garrett361 please review

File: docs/model-dev-guide/dtrain/dtrain-introduction.rst [hunk 20]
```diff
@@ -73,11 +65,12 @@ you have scaled your ``global_batch_size`` proportionally with ``slots_per_trial
...
-number (e.g., 1000). This setting results in communicating updates once every 1000 batches.
-Comparing throughput with ``aggregation_frequency`` of 1 vs. ``aggregation_frequency`` of 1000 will
-demonstrate the communication overhead. If you do observe significant communication overhead, refer
-to :ref:`multi-gpu-training` for guidance on how to optimize communication.
+``optimizations.aggregation_frequency`` in the :ref:`experiment configuration
+<experiment-config-reference>` to a very large number (e.g., 1000). This setting results in
+communicating updates once every 1000 batches. Comparing throughput with an
+``aggregation_frequency`` of 1 vs. an ``aggregation_frequency`` of 1000 will demonstrate the
+communication overhead. If you do observe significant communication overhead, refer to
+:ref:`multi-gpu-training-implement-adv-optimizations` for guidance on how to optimize communication.
```
Comment (tara-det-ai) [inline docs/model-dev-guide/dtrain/dtrain-introduction.rst:73]: @garrett361 please review

File: docs/model-dev-guide/dtrain/dtrain-implement.rst (hunk 19 above)
Comment (garrett361) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:84]: I think something like the following might be a bit more precise: "The `global_batch_size` field will be automatically respected by the Trial APIs.  Core API users must reference this field explicitly and organize their code to respect its value in order to use this hyperparameter."

**Review by garrett361 — APPROVED** [review]
LGTM, just one small comment.

File: docs/model-dev-guide/dtrain/dtrain-introduction.rst (hunk 20 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-introduction.rst:73]: resolved, i double checked the link the content to be sure one was supposed to reference the other

File: docs/model-dev-guide/dtrain/dtrain-implement.rst (hunk 19 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/dtrain/dtrain-implement.rst:84]: thanks. resolved.
